src/libcore/intrinsics.rs

```diff
@@ -278,17 +278,157 @@ extern "rust-intrinsic" {
+    /// {
+    ///     let ptr = &mut 0;
+    ///     let val_transmuted = mem::transmute::<&mut i32, &mut u32>(ptr);
+    ///     // Now let's put together `as` and reborrowing
+    ///     let val_casts = &mut *(ptr as *mut i32 as *mut u32);
+    /// }
+    ///
+    /// // Turning an `&str` into an `&[u8]`
+    /// {
+    ///     // this is not a good way to do this.
+    ///     let slice = unsafe { mem::transmute::<&str, &[u8]>("Rust") };
+    ///     assert_eq!(slice, [82, 117, 115, 116]);
+    ///     // You could use `str::as_bytes`
+    ///     let slice = "Rust".as_bytes();
+    ///     assert_eq!(slice, [82, 117, 115, 116]);
+    ///     // Or, just use a byte string, if you have control over the string
+    ///     // literal
+    ///     assert_eq!(b"Rust", [82, 117, 116, 116]);
+    /// }
+    ///
+    /// // Turning a Vec<&T> into a Vec<Option<&T>>
+    /// {
+    ///     let store = [0, 1, 2, 3];
+    ///     let v_orig = store.iter().collect::<Vec<&i32>>();
+    ///     // Using transmute; Undefined Behavior
+    ///     let v_transmuted = mem::transmute::<Vec<&i32>, Vec<Option<&i32>>>(
+    ///         v_orig.clone());
+    ///     // The suggested, safe way
+    ///     let v_collected = v_orig.clone()
+    ///                             .into_iter()
+    ///                             .map(|r| Some(r))
+    ///                             .collect::<Vec<Option<&i32>>>();
+    ///     // The no-copy, unsafe way, still using transmute, but not UB
+    ///     // This is equivalent to the original, but safer, and reuses the
+    ///     // same Vec internals. Therefore the new inner type must have the
+    ///     // exact same size, and the same or lesser alignment, as the old
+    ///     // type. The same caveats exist for this method as transmute, for
+    ///     // the original inner type (`&i32`) to the converted inner type
+    ///     // (`Option<&i32>`), so read the nomicon page linked above.
+    ///     let v_no_copy = Vec::from_raw_parts(v_orig.as_mut_ptr(),
+    ///                                         v_orig.len(),
+    ///                                         v_orig.capacity());
+    ///     mem::forget(v_orig);
+    /// }
+    ///
+    ///
+    /// // Copying an `&mut T` to reslice:
+    /// {
+    ///     fn split_at_mut_transmute<T>(slice: &mut [T], index: usize)
+    ///                                  -> (&mut [T], &mut [T]) {
+    ///         let len = slice.len();
+    ///         assert!(index < len);
+    ///         let slice2 = mem::transmute::<&mut [T], &mut [T]>(slice);
+    ///         (slice[0..index], slice2[index..len])
+    ///     }
+    ///     // Again, use `as` and reborrowing
+    ///     fn split_at_mut_casts<T>(slice: &mut [T], index: usize)
+    ///                              -> (&mut [T], &mut [T]) {
+    ///         let len = slice.len();
+    ///         assert!(index < len);
+    ///         let slice2 = &mut *(slice as *mut [T]); // actually typesafe!
```

Comment:
`slice` and `slice2` are aliased here so is this really safe?

This isn't how [`split_at_mut`](https://github.com/rust-lang/rust/blob/master/src/libcore/slice.rs#L339-L349) is actually implemented so I'm not sure this is a good example and the advice should be just to use the method in `std`.